File: adversarial-harness/python/policy_engine/timing_guard.py
"""Timing side-channel probe detection."""

from __future__ import annotations

import os
import re
import time
from typing import Optional

from .types import CallContext, PolicyDecision

TIMING_PROBE = [
    re.compile(r"\b(?:sleep|benchmark|pg_sleep|waitfor\s+delay)\s*\(", re.I),
    re.compile(r"\bif\s*\(\s*(?:ascii|ord|substring)", re.I),
    re.compile(r"\b(?:timing|time[- ]?based)\s+(?:attack|oracle)", re.I),
]

_probe_counts: dict[str, tuple[int, float]] = {}
MAX_PROBES = int(os.environ.get("MCP_GUARDIAN_MAX_TIMING_PROBES_PER_MIN", "8"))


def reset_timing_probe_counters() -> None:
    _probe_counts.clear()


def evaluate_timing_guard(ctx: CallContext) -> Optional[PolicyDecision]:
    import json

    blob = json.dumps(ctx.arguments or {}, ensure_ascii=False)
    if not any(p.search(blob) for p in TIMING_PROBE):
        return None
    tenant = ctx.tenant_id or "default"
    sub = ctx.agent_identity.sub if ctx.agent_identity else "anon"
    key = f"timing:{tenant}:{ctx.server_name}:{sub}"
    now = time.time()
    count, reset = _probe_counts.get(key, (0, now + 60))
    if now > reset:
        count, reset = 1, now + 60
    else:
        count += 1
    _probe_counts[key] = (count, reset)
    if count > MAX_PROBES:
        return PolicyDecision("block", "timing-probe-rate-limit", "Timing probe rate exceeded")
    return PolicyDecision("block", "timing-side-channel-guard", "Timing side-channel probe in arguments")

File: src/policy/timing-guard.ts
/**
 * Timing side-channel mitigation — blocks timing oracle probes in tool arguments.
 */
import { LRUCache } from 'lru-cache';
import type { CallContext, PolicyDecision } from './policy-types.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import { deobfuscateRecursive } from '../utils/payload-normalizer.js';

const TIMING_PROBE_PATTERNS: RegExp[] = [
  /\b(?:sleep|benchmark|pg_sleep|waitfor\s+delay|dbms_lock\.sleep)\s*\(/i,
  /\bif\s*\(\s*(?:ascii|ord|substring|substr)\s*\(/i,
  /\b(?:case\s+when|elt\s*\()\s+.*\b(?:sleep|benchmark|waitfor)/i,
  /\b(?:timing|time[- ]?based)\s+(?:attack|oracle|injection)/i,
  /\b(?:measure|detect)\s+(?:response\s+)?time\s+(?:of|for)\s+(?:user|login|password)/i,
  /\b(?:valid|invalid)\s+username\b.*\b(?:timing|delay|sleep)/i,
  /\busername\s+exists\b.*\b(?:time|delay|benchmark)/i,
];

const PROBE_WINDOW_MS = 60_000;
const MAX_TIMING_PROBES_PER_SESSION = parseInt(
  process.env['MCP_GUARDIAN_MAX_TIMING_PROBES_PER_MIN'] ?? '8',
  10,
);

const probeCounters = new LRUCache<string, { count: number; resetAt: number }>({
  max: 20_000,
  ttl: PROBE_WINDOW_MS,
});

function probeSessionKey(ctx: CallContext): string {
  const tenant = ctx.tenantId || process.env['GUARDIAN_TENANT_ID'] || 'default';
  const sub = ctx.agentIdentity?.sub || ctx.agentIdentity?.clientId || 'anon';
  return `timing:${tenant}:${ctx.serverName}:${sub}`;
}

function blobLooksLikeTimingProbe(blob: string): boolean {
  return TIMING_PROBE_PATTERNS.some((p) => p.test(blob));
}

export function resetTimingProbeCounters(): void {
  probeCounters.clear();
}

export function evaluateTimingGuard(ctx: CallContext): PolicyDecision | null {
  const args = ctx.arguments ?? {};
  const blob = walkStringLeaves(args)
    .map((l) => deobfuscateRecursive(l.value))
    .join('\n');

  if (!blob.trim()) return null;

  if (!blobLooksLikeTimingProbe(blob)) return null;

  const key = probeSessionKey(ctx);
  const now = Date.now();
  let counter = probeCounters.get(key);
  if (!counter || now > counter.resetAt) {
    counter = { count: 1, resetAt: now + PROBE_WINDOW_MS };
  } else {
    counter.count++;
  }
  probeCounters.set(key, counter);

  if (counter.count > MAX_TIMING_PROBES_PER_SESSION) {
    return {
      action: 'block',
      rule: 'timing-probe-rate-limit',
      reason: `Timing oracle probe rate exceeded (${counter.count}/${MAX_TIMING_PROBES_PER_SESSION} per minute)`,
    };
  }

  return {
    action: 'block',
    rule: 'timing-side-channel-guard',
    reason: 'Timing-based side-channel probe pattern in tool arguments',
  };
}
